jax/_src/numpy/lax_numpy.py

```diff
@@ -3527,6 +3527,94 @@ def _rewriting_take(arr, idx, indices_are_sorted=False, unique_indices=False,
   return _gather(arr, treedef, static_idx, dynamic_idx, indices_are_sorted,
                  unique_indices, mode, fill_value)
 
+invert_permutation = lambda seq: sorted(range(len(seq)), key=seq.__getitem__)
+def _get_order_with_axes_shift_forward(ndim: int, axes: tuple, inverse: bool=False, forward: bool=True):
+  other_axes = tuple([i for i in range(ndim) if i not in axes])
+  fwd_order = axes + other_axes if forward else other_axes + axes
+  order = fwd_order if not inverse else invert_permutation(fwd_order)
+  return order
+
+def shift_axes_forward(operand: core.Tracer, axes: tuple, inverse: bool=False,
+                       forward: bool=True):
+  """Shifts the tuple of axes to the front of array"""
+  order = _get_order_with_axes_shift_forward(len(operand.shape), axes, inverse, forward)
+  return lax.transpose(operand, order)
+
+def shift_axes_forward_tuple(operand: tuple, axes: tuple, inverse: bool=False,
+                       forward: bool=True):
+  """Shifts the tuple of axes to the front of tuple"""
+  order = _get_order_with_axes_shift_forward(len(operand), axes, inverse, forward)
+  return [operand[k] for k in order]
+
+def get_slice_indices(start_indices: tuple, gather_slice_shape: tuple, array_shape: tuple):
+  """
+  Calculates the slice shapes for PROMISE_IN_BOUNDS
+  """
+  # inbuilt python functions like min are overwritten by np.min so we explicitly define
+  _add = lambda x, y: x + y
+  _min = lambda x, y: x if x < y else y
+  _max = lambda x, y: x if x > y else y
+  # shift if start_indices are lower-out-of-bounds
+  start_indices = tuple(map(lambda i: _max(0, i), start_indices))
+  # calculate limit_indices
+  limit_indices = tuple(map(_add, start_indices, gather_slice_shape))
+  # shift if limit_indices are upper-out-of-bounds
+  bounds_shift = tuple(map(lambda x, y: _min(0, x-y), array_shape, limit_indices))
+  start_indices, limit_indices = [tuple(map(_add, indices, bounds_shift)) for indices in [start_indices, limit_indices]]
+  return start_indices, limit_indices
+
+def _lower_gather_scatter(arr, indexer, update=None, op=None):
```

Comment:
This now has ugly semantics to support both scatter and gather as they share alot in common. Not sure of best way to proceed here.

Reply:
Splitting methods. Also will abandon the reorder of indices approach in favour of a concat as it removes the tricky ordering of indices back and forth